flax/experimental/nnx/nnx/nn/attention.py

```diff
@@ -491,39 +491,44 @@ def __call__(
     # During fast autoregressive decoding, we feed one position at a time,
     # and cache the keys and values step by step.
     if self.decode:
-      (
-        *batch_dims,
-        max_length,
-        num_heads,
-        depth_per_head,
-      ) = self.cached_key.shape
-      # shape check of cached keys against query input
-      expected_shape = tuple(batch_dims) + (1, num_heads, depth_per_head)
-      if expected_shape != query.shape:
-        raise ValueError(
-          'Autoregressive cache shape error, '
-          'expected query shape %s instead got %s.'
-          % (expected_shape, query.shape)
+      is_initialized = hasattr(self, 'cached_key')
+      if not is_initialized:
+        self.init_cache(key, value)
```

Comment:
To initialize cache we are proposing this pattern:
https://github.com/google/flax/blob/04642eb344b22e76aa91194321c938b3e32a387c/flax/experimental/nnx/examples/lm1b/train.py#L254-L259

Reply:
The trailing dimensions for the `cached_key` in NNX are [`(..., self.num_heads, self.out_features)`](https://github.com/google/flax/blob/04642eb344b22e76aa91194321c938b3e32a387c/flax/experimental/nnx/nnx/nn/attention.py#L569) whereas in Linen, the shape is the [`key` shape](https://github.com/google/flax/blob/04642eb344b22e76aa91194321c938b3e32a387c/flax/linen/attention.py#L471) itself `(..., self.num_heads, qkv_features // self.num_heads)`, which is derived from [`DenseGeneral`](https://github.com/google/flax/blob/04642eb344b22e76aa91194321c938b3e32a387c/flax/linen/attention.py#L443). Is there a reason for this discrepancy?

Reply:
Thanks for catching this! Its a bug, probably due the same misuse of `out_features` you pointed out in another issue.